src/fdc/MSXtar.cc

```diff
@@ -50,22 +50,56 @@ static constexpr uint8_t T_MSX_ARC  = 0x20; // Archive bit
 // For details, read http://home.teleport.com/~brainy/lfn.htm
 static constexpr uint8_t T_MSX_LFN  = 0x0F; // LFN entry (long files names)
 
```

Comment:
I 100% agree with the idea of the patch. But I do not like the API of the new `Cluster` class, nor the implementation.

I would suggest the following:
```
class Cluster
{
private:
	explicit Cluster(unsigned value) : val(value) {}

public:
    [[nodiscard]] static Cluster FREE()         { return Cluster(FREE_FAT); }
	[[nodiscard]] static Cluster END_OF_CHAIN() { return Cluster(EOF_FAT); }

	[[nodiscard]] static Cluster fromIndex(unsigned index) {
        assert(index + FIRST_CLUSTER < BAD_FAT);
        return Cluster(index + FIRST_CLUSTER);
    }
	[[nodiscard]] static Cluster fromValue(unsigned value, unsigned clusterCount) {
        assert(clusterCount < BAD_FAT);
        if (value == 0) {
		    return FREE();
	    } else if (value < FIRST_CLUSTER || value >= clusterCount) {
		    return END_OF_CHAIN();
	    } else {
            return Cluster(value);
        }
    }

	[[nodiscard]] bool isValidIndex() const {
        return (FIRST_CLUSTER <= val) && (val < BAD_FAT);
    }
	[[nodiscard]] unsigned index() const {
        assert(isValidIndex());
        return val - FIRST_CLUSTER;
    }

	[[nodiscard]] unsigned value() const {
        return val;
    }

	[[nodiscard]] bool operator==(const Cluster& other) const = default;

private:
	unsigned val;
};
```

Most important items:
* There are no public constructors. Instead a `Cluster` object must be obtained either via `Cluster::fromIndex()` or `Cluster::fromValue()`. (Or via `Cluster::FREE()` or `Cluster::END_OF_CHAIN()`).
* Internally the class stores a single `unsigned` value. There's an invariant that this is either a valid index or FREE_FAT or EOF_FAT. (So this makes the separate `type` enum obsolete).
* There are only two queries on this class: `index()` and `value()`. The former has a precondition that `isValidIndex()` must be true (OK, this is the 3rd query method). The latter simply returns the internal value.
* operator== can now simply compare the internal value.

I'll give a few examples below for how this can replace your implementation.

Reply:
So, the idea behind the Cluster implementation was to implement a (tagged) typed union, which represents all values completely independently in the most natural way. Since C++ doesn’t really have such a built in concept (well, could’ve used polymorphism, but I wanted to keep it simple), it needs to be explicitly tagged with an enum, where the first two are atomic types and the last has a value.

(In Haskell it would be `data ClusterNumber a : Free | EndOfChain | Cluster a`.)

What I do like about this implementation is that it has the special values and the integer completely separated. The fields have their own meaning and only that. No magic values. This way, there is no way the value can somehow overflow and mean something completely different, which gives it a level of safeness that asserts can’t provide. Which addresses an aspect of the code that I didn’t like before this change.

After I got it working, I thought about two possible ways to optimise:

1. Use bit fields to store the value in the space of a single integer

I played around a bit in compiler explorer and it did not really result in especially nice assembly code, it was not more efficient than the original. So that’s just a size optimisation, and hardly even that because the struct is passed around in registers anyway.

2. Combine the value in a single integer as you suggest to do.

I felt it was an optimisation that would optimise the comparison but not much else, and it would make the code more complicated to read. More importantly it would remove that property of safeness from overflowing.

Note that for `FREE` and `EOC` I would use magic values far away from the cluster index (like `~0` and `~1`), so that makes the code more complicated than yours. Because FAT12 and FAT16 use different cluster numbers for EOC, while this object is oblivious of the FAT type, it’s got its own magic values. These far-away values would perhaps also reduce the chance of an overflow changing the type (though it’s still easy for an underflow).

I used std::optional() because it puts a stronger pressure on the programmer to check the presence of a return value before using it, even though it didn’t gel super well with me syntactically, that much is true. But semantically, the meaning is, the value exists, or it does not, because it’s a different type.

About your suggestion specifically, note that I understand the line of thought, as mentioned I had it myself as well. However. I think it removes aspects of the design intent of this type, without regard, for the sake of optimisation only. Without that, is there still a benefit in having this abstraction? I feel it might be clearer and simpler the way it was before.

p.s. Since asserts are compiled out (ew) in builds used by most people, honestly I don’t trust them one bit.

Reply:
In my mind this type is similar to `std::optional` in terms of its semantics as a tagged type union.

Optional represents either the value or bottom (aka "nothing", "nil", "undefined", "error", "uncomputable").

This type represents either Cluster(index) or Free or EndOfChain.

...and I just found out about [`std::variant`](https://en.cppreference.com/w/cpp/utility/variant), which would probably be a better way to represent this.

Reply:
Yes, the idiomatic way to implement a tagged-union in c++ is via `std::variant`.
Here's an example of how to do that:   https://godbolt.org/z/MsGvKsMYe
Under the hood a std::variant is much like the code you implemented manually (internally it has a small integer to indicate the active alternative plus the (union of the) members of those different alternatives).

I can see the appeal of why you want to model it like this.
Personally I wouldn't do it like this, because the thing you want to model really is an integer (with special meaning for certain (ranges of) values). But you decide, either is fine for me.
I must also admit that I haven't yet thought about how this class will change with future FAT16 support.

(Maybe one small thing to try to convince you: if in my implementation you store the member as a `Endian::L16` then you could use this class directly in `MSXDirEntry`.)

Reply:
UPDATE: I read your FAT16 follow-up patches. I can see now how the tagged-union approach makes more code independent from FAT12 vs FAT16 (the decision for how to represent "end-of-chain" can be postponed). So that's a bonus.

I'm personally undecided which approach I like better, So again, I'll let you decide.